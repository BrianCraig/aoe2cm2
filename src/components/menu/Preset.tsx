import * as React from "react";
import {Link} from "react-router-dom";

class Preset extends React.Component<object, object> {
    public render() {
        return (
            <div>
                <p>Description of a single preset</p>
                <Link to='/'>Go to index</Link>
            </div>
        );
    }
}

export default Preset;