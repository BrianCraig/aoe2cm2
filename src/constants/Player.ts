enum Player {
    HOST = 'HOST',
    GUEST = 'GUEST',
    NONE = 'NONE'
}

export default Player;