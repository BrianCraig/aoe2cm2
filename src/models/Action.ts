enum Action {
    PICK = 'PICK',
    EXCLUSIVE_PICK = 'EXCLUSIVE_PICK',
    GLOBAL_PICK = 'GLOBAL_PICK',
    HIDDEN_PICK = 'HIDDEN_PICK',
    HIDDEN_EXCLUSIVE_PICK = 'HIDDEN_EXCLUSIVE_PICK',
    BAN = 'BAN',
    EXCLUSIVE_BAN = 'EXCLUSIVE_BAN',
    GLOBAL_BAN = 'GLOBAL_BAN',
    HIDDEN_BAN = 'HIDDEN_BAN',
    HIDDEN_EXCLUSIVE_BAN = 'HIDDEN_EXCLUSIVE_BAN',
    HIDDEN_GLOBAL_BAN = 'HIDDEN_GLOBAL_BAN',
    SNIPE = 'SNIPE',
    HIDDEN_SNIPE = 'HIDDEN_SNIPE',
    
    // REVEAL_PICKS = 'REVEAL_PICKS',
    // REVEAL_BANS = 'REVEAL_BANS',
    // REVEAL_SNIPES = 'REVEAL_SNIPES',
    REVEAL_ALL = 'REVEAL_ALL'
}

export default Action;