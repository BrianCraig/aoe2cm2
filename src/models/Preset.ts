import Civilisation from "./Civilisation";
import Turn from "./Turn";
import {CivilisationEncoder} from "./CivilisationEncoder";
import {Assert} from "./Assert";

class Preset {

    public static readonly EMPTY: Preset = new Preset('', [], []);

    public static readonly SAMPLE: Preset = new Preset('Default Preset', Civilisation.ALL, [
        Turn.HOST_GLOBAL_BAN,
        Turn.GUEST_GLOBAL_BAN,
        Turn.HOST_HIDDEN_BAN,
        Turn.GUEST_HIDDEN_BAN,
        Turn.REVEAL_ALL,
        Turn.HOST_GLOBAL_PICK,
        Turn.GUEST_GLOBAL_PICK,
        Turn.HOST_PICK,
        Turn.GUEST_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.REVEAL_ALL,
        Turn.HOST_HIDDEN_SNIPE,
        Turn.GUEST_HIDDEN_SNIPE,
        Turn.REVEAL_ALL
    ]);

    public static readonly SIMPLE: Preset = new Preset('Simple Preset', Civilisation.ALL, [
        Turn.HOST_NONEXCLUSIVE_BAN,
        Turn.GUEST_NONEXCLUSIVE_BAN,
        Turn.GUEST_NONEXCLUSIVE_PICK,
        Turn.HOST_NONEXCLUSIVE_PICK
    ]);

    public static readonly HIDDEN_1V1: Preset = new Preset('Hidden 1v1', Civilisation.ALL, [
        Turn.HOST_HIDDEN_BAN,
        Turn.GUEST_HIDDEN_BAN,
        Turn.GUEST_HIDDEN_BAN,
        Turn.HOST_HIDDEN_BAN,
        Turn.HOST_HIDDEN_BAN,
        Turn.GUEST_HIDDEN_BAN,
        Turn.REVEAL_ALL,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.REVEAL_ALL
    ]);

    public static readonly HIDDEN_2V2: Preset = new Preset('Hidden 2v2', Civilisation.ALL, [
        Turn.HOST_HIDDEN_BAN,
        Turn.GUEST_HIDDEN_BAN,
        Turn.HOST_HIDDEN_BAN,
        Turn.GUEST_HIDDEN_BAN,
        Turn.HOST_HIDDEN_BAN,
        Turn.GUEST_HIDDEN_BAN,
        Turn.REVEAL_ALL,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.REVEAL_ALL,
        Turn.GUEST_HIDDEN_BAN,
        Turn.HOST_HIDDEN_BAN,
        Turn.REVEAL_ALL,
        Turn.GUEST_HIDDEN_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.REVEAL_ALL,
    ]);

    public static readonly HIDDEN_3V3: Preset = new Preset('Hidden 3v3', Civilisation.ALL, [
        Turn.HOST_HIDDEN_BAN,
        Turn.GUEST_HIDDEN_BAN,
        Turn.HOST_HIDDEN_BAN,
        Turn.GUEST_HIDDEN_BAN,
        Turn.REVEAL_ALL,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.REVEAL_ALL,
        Turn.HOST_HIDDEN_BAN,
        Turn.GUEST_HIDDEN_BAN,
        Turn.REVEAL_ALL,
        Turn.GUEST_HIDDEN_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.REVEAL_ALL,
        Turn.GUEST_HIDDEN_BAN,
        Turn.HOST_HIDDEN_BAN,
        Turn.REVEAL_ALL,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.REVEAL_ALL,
    ]);

    public static readonly HIDDEN_4V4: Preset = new Preset('Hidden 4v4', Civilisation.ALL, [
        Turn.HOST_HIDDEN_BAN,
        Turn.GUEST_HIDDEN_BAN,
        Turn.HOST_HIDDEN_BAN,
        Turn.GUEST_HIDDEN_BAN,
        Turn.REVEAL_ALL,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.REVEAL_ALL,
        Turn.HOST_HIDDEN_BAN,
        Turn.GUEST_HIDDEN_BAN,
        Turn.REVEAL_ALL,
        Turn.GUEST_HIDDEN_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.REVEAL_ALL,
        Turn.GUEST_HIDDEN_BAN,
        Turn.HOST_HIDDEN_BAN,
        Turn.REVEAL_ALL,
        Turn.GUEST_HIDDEN_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.REVEAL_ALL,
    ]);

    public readonly name: string;
    public readonly encodedCivilisations: string;
    public readonly turns: Turn[];

    constructor(name: string, civilisations: Civilisation[], turns: Turn[] = []) {
        this.name = name;
        this.encodedCivilisations = CivilisationEncoder.encodeCivilisationArray(civilisations);
        this.turns = turns;
    }

    public static fromPojo(preset: { name: string, encodedCivilisations: string, turns: Turn[] } | undefined): Preset | undefined {
        if (preset === undefined) {
            return undefined;
        }
        Assert.isString(preset.name);
        Assert.isString(preset.encodedCivilisations);
        return new Preset(preset.name, CivilisationEncoder.decodeCivilisationArray(preset.encodedCivilisations), Turn.fromPojoArray(preset.turns));
    }

    public addTurn(turn: Turn) {
        this.turns.push(turn);
    }

    get civilisations(): Civilisation[] {
        return CivilisationEncoder.decodeCivilisationArray(this.encodedCivilisations);
    }
}

export default Preset;