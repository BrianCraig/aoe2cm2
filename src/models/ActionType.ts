import Action from "./Action";

enum ActionType {
    PICK = "pick",
    BAN = "ban",
    SNIPE = "snipe",
    REVEAL = "reveal",
    NOTHING = "nothing"
}

export function actionTypeFromAction(action: Action) {
    switch (action) {
        case Action.PICK:
        case Action.EXCLUSIVE_PICK:
        case Action.GLOBAL_PICK:
        case Action.HIDDEN_PICK:
        case Action.HIDDEN_EXCLUSIVE_PICK:
            return ActionType.PICK;
        case Action.BAN:
        case Action.EXCLUSIVE_BAN:
        case Action.GLOBAL_BAN:
        case Action.HIDDEN_BAN:
        case Action.HIDDEN_EXCLUSIVE_BAN:
        case Action.HIDDEN_GLOBAL_BAN:
            return ActionType.BAN;
        case Action.SNIPE:
        case Action.HIDDEN_SNIPE:
            return ActionType.SNIPE;
        // case Action.REVEAL_PICKS:
        // case Action.REVEAL_BANS:
        // case Action.REVEAL_SNIPES:
        case Action.REVEAL_ALL:
            return ActionType.REVEAL;
    }
}

export default ActionType;