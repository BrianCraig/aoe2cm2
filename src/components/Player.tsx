import * as React from 'react';
import ActionType from "../models/ActionType";
import {default as ModelPlayer} from '../models/Player';
import {default as ModelPreset} from '../models/Preset';
import '../pure-min.css'
import '../style2.css'
import CivPanel from "../containers/CivPanel";
import Civilisation from "../models/Civilisation";
import {DraftEvent} from "../models/DraftEvent";
import PlayerEvent from "../models/PlayerEvent";

interface IProps {
    preset: ModelPreset;
    player: ModelPlayer;
    name: string;
    events?: DraftEvent[];
}

interface IState {
    bannedCivs: Civilisation[];
    pickedCivs: Civilisation[];
    snipedCivs: Civilisation[];
}

class Player extends React.Component<IProps, IState> {

    constructor(props: IProps) {
        super(props);
        this.state = eventsToState(props.events, props.player);
    }

    public componentWillReceiveProps(nextProps: Readonly<IProps>, nextContext: any): void {
        console.log("Propping up", nextProps);
        this.setState(eventsToState(nextProps.events, nextProps.player));
    }

    public render() {
        const playerId: string = 'player-' + this.props.player.toString().toLowerCase();

        const pickPanels = this.props.preset.turns.filter(turn => {
            return turn.player === this.props.player && turn.action.toString().includes('PICK');
        }).map((turn, index) => {
            let pickedCiv;
            if (this.state.pickedCivs.length > index) {
                pickedCiv = this.state.pickedCivs[index];
            }
            return React.createElement(CivPanel, {actionType: ActionType.PICK, active: false, civilisation: pickedCiv});
        });
        const banPanels = this.props.preset.turns.filter(turn => {
            return turn.player === this.props.player && turn.action.toString().includes('BAN');
        }).map((turn, index) => {
            let bannedCiv;
            if (this.state.bannedCivs.length > index) {
                bannedCiv = this.state.bannedCivs[index];
            }
            return React.createElement(CivPanel, {actionType: ActionType.BAN, active: false, civilisation: bannedCiv});
        });

        return (
            <div className="pure-u-1-2">
                <div id={playerId} className="double-outer-border">
                    <div className="double-inner-border">
                        <div className="player">
                            <div className="head-text">
                                {this.props.player}
                            </div>
                            <div className="player-head">
                                <div className="player-name">{this.props.name}</div>
                            </div>
                            <div className="chosen">
                                <div className="head-text">Picks</div>
                                <div className="picks">
                                    {pickPanels}
                                </div>

                                <div className="head-text">Bans</div>
                                <div className="bans">
                                    {banPanels}
                                </div>
                            </div>

                        </div>
                    </div>
                </div>
            </div>
        );
    }
}


function eventsToState(events: DraftEvent[] | undefined, player: ModelPlayer): IState {
    if (events === undefined) {
        return {bannedCivs: [], pickedCivs: [], snipedCivs: []};
    } else {
        events = events as DraftEvent[];
    }
    const playerEvents: PlayerEvent[] = events
        .filter(e => e instanceof PlayerEvent)
        .map(e => e as PlayerEvent)
        .filter(e => e.player === player);
    const picks: Civilisation[] = [];
    const bans: Civilisation[] = [];
    const snipes: Civilisation[] = [];
    for (const event of playerEvents) {
        switch (event.actionType) {
            case ActionType.PICK:
                picks.push(event.civilisation);
                break;
            case ActionType.BAN:
                bans.push(event.civilisation);
                break;
            case ActionType.SNIPE:
                snipes.push(event.civilisation);
                break;
        }
    }
    return {bannedCivs: bans, pickedCivs: picks, snipedCivs: snipes};
}

export default Player;
