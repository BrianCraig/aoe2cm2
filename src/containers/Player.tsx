import {IStoreState} from '../types';
import {connect} from 'react-redux';
import Player from "../components/Player";


export function mapStateToProps(state: IStoreState) {
    return {
        events: state.events,
        nextAction: state.nextAction
    };
}

export function mapDispatchToProps() {
    return {};
}

export default connect(mapStateToProps, mapDispatchToProps)(Player);