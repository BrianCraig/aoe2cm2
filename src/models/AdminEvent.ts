import Player from "./Player";
import Action from "./Action";

class AdminEvent {
    public readonly player: Player;
    public readonly action: Action;

    constructor(player: Player, action: Action) {
        this.player = player;
        this.action = action;
    }
}

export default AdminEvent;
