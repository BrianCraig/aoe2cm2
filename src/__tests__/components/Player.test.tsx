import {shallow} from "enzyme";
import * as React from "react";
import {default as ModelPlayer} from "../../models/Player";
import Preset from "../../models/Preset";
import Player from "../../components/draft/PlayerDraftState";

it('Player renders correctly', () => {
    const component = shallow(<Player preset={Preset.SAMPLE} name={"Host Player Name"} player={ModelPlayer.HOST}/>);
    expect(component).toMatchSnapshot();
});
