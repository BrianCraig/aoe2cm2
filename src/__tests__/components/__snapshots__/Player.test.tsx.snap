// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Player renders correctly 1`] = `
<Player
  i18n={Object {}}
  name="Host Player Name"
  player="HOST"
  preset={
    Preset {
      "encodedCivilisations": "7fffffff",
      "name": "Default Preset",
      "turns": Array [
        Turn {
          "action": "BAN",
          "exclusivity": "GLOBAL",
          "hidden": false,
          "parallel": false,
          "player": "HOST",
        },
        Turn {
          "action": "BAN",
          "exclusivity": "GLOBAL",
          "hidden": false,
          "parallel": false,
          "player": "GUEST",
        },
        Turn {
          "action": "BAN",
          "exclusivity": "NONEXCLUSIVE",
          "hidden": true,
          "parallel": false,
          "player": "HOST",
        },
        Turn {
          "action": "BAN",
          "exclusivity": "NONEXCLUSIVE",
          "hidden": true,
          "parallel": false,
          "player": "GUEST",
        },
        Turn {
          "action": "REVEAL_ALL",
          "exclusivity": "GLOBAL",
          "hidden": false,
          "parallel": false,
          "player": "NONE",
        },
        Turn {
          "action": "PICK",
          "exclusivity": "GLOBAL",
          "hidden": false,
          "parallel": false,
          "player": "HOST",
        },
        Turn {
          "action": "PICK",
          "exclusivity": "GLOBAL",
          "hidden": false,
          "parallel": false,
          "player": "GUEST",
        },
        Turn {
          "action": "PICK",
          "exclusivity": "EXCLUSIVE",
          "hidden": false,
          "parallel": false,
          "player": "HOST",
        },
        Turn {
          "action": "PICK",
          "exclusivity": "EXCLUSIVE",
          "hidden": false,
          "parallel": false,
          "player": "GUEST",
        },
        Turn {
          "action": "PICK",
          "exclusivity": "NONEXCLUSIVE",
          "hidden": true,
          "parallel": false,
          "player": "HOST",
        },
        Turn {
          "action": "PICK",
          "exclusivity": "NONEXCLUSIVE",
          "hidden": true,
          "parallel": false,
          "player": "GUEST",
        },
        Turn {
          "action": "PICK",
          "exclusivity": "NONEXCLUSIVE",
          "hidden": true,
          "parallel": false,
          "player": "HOST",
        },
        Turn {
          "action": "PICK",
          "exclusivity": "NONEXCLUSIVE",
          "hidden": true,
          "parallel": false,
          "player": "GUEST",
        },
        Turn {
          "action": "PICK",
          "exclusivity": "NONEXCLUSIVE",
          "hidden": true,
          "parallel": false,
          "player": "HOST",
        },
        Turn {
          "action": "PICK",
          "exclusivity": "NONEXCLUSIVE",
          "hidden": true,
          "parallel": false,
          "player": "GUEST",
        },
        Turn {
          "action": "PICK",
          "exclusivity": "NONEXCLUSIVE",
          "hidden": true,
          "parallel": false,
          "player": "HOST",
        },
        Turn {
          "action": "PICK",
          "exclusivity": "NONEXCLUSIVE",
          "hidden": true,
          "parallel": false,
          "player": "GUEST",
        },
        Turn {
          "action": "REVEAL_ALL",
          "exclusivity": "GLOBAL",
          "hidden": false,
          "parallel": false,
          "player": "NONE",
        },
        Turn {
          "action": "SNIPE",
          "exclusivity": "NONEXCLUSIVE",
          "hidden": true,
          "parallel": false,
          "player": "HOST",
        },
        Turn {
          "action": "SNIPE",
          "exclusivity": "NONEXCLUSIVE",
          "hidden": true,
          "parallel": false,
          "player": "GUEST",
        },
        Turn {
          "action": "REVEAL_ALL",
          "exclusivity": "GLOBAL",
          "hidden": false,
          "parallel": false,
          "player": "NONE",
        },
      ],
    }
  }
  t={[Function]}
  tReady={true}
/>
`;
